File: src/blackboxd/providers/anthropic.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from blackboxd.capture import capture_llm_call
from blackboxd.models import TokenUsage
from blackboxd.storage.base import Storage


def _utc_now() -> datetime:
    return datetime.now(timezone.utc)


def _extract_usage(response: Any) -> TokenUsage | None:
    usage = getattr(response, "usage", None)
    if usage is None:
        return None
    return TokenUsage(
        input_tokens=getattr(usage, "input_tokens", None),
        output_tokens=getattr(usage, "output_tokens", None),
    )


class _MessagesAPI:
    def __init__(self, client: Any, *, provider: str, storage: Storage | str | None) -> None:
        self._client = client
        self._provider = provider
        self._storage = storage

    def create(self, **kwargs):
        model = kwargs.get("model")
        started_at = _utc_now()
        try:
            response = self._client.messages.create(**kwargs)
        except Exception as exc:
            capture_llm_call(
                name="anthropic.messages.create",
                provider=self._provider,
                model=model,
                prompt=kwargs,
                started_at=started_at,
                ended_at=_utc_now(),
                error=exc,
                storage=self._storage,
            )
            raise
        capture_llm_call(
            name="anthropic.messages.create",
            provider=self._provider,
            model=model,
            prompt=kwargs,
            started_at=started_at,
            ended_at=_utc_now(),
            response=response,
            usage=_extract_usage(response),
            storage=self._storage,
        )
        return response


class _AsyncMessagesAPI(_MessagesAPI):
    async def create(self, **kwargs):
        model = kwargs.get("model")
        started_at = _utc_now()
        try:
            response = await self._client.messages.create(**kwargs)
        except Exception as exc:
            capture_llm_call(
                name="anthropic.messages.create",
                provider=self._provider,
                model=model,
                prompt=kwargs,
                started_at=started_at,
                ended_at=_utc_now(),
                error=exc,
                storage=self._storage,
            )
            raise
        capture_llm_call(
            name="anthropic.messages.create",
            provider=self._provider,
            model=model,
            prompt=kwargs,
            started_at=started_at,
            ended_at=_utc_now(),
            response=response,
            usage=_extract_usage(response),
            storage=self._storage,
        )
        return response


class _BaseAnthropicWrapper:
    def __init__(self, client: Any, *, provider: str = "anthropic", storage: Storage | str | None = None) -> None:
        self._client = client
        self.messages = self._messages_cls(client, provider=provider, storage=storage)

    def __getattr__(self, item: str) -> Any:
        return getattr(self._client, item)


class Anthropic(_BaseAnthropicWrapper):
    _messages_cls = _MessagesAPI

    def __init__(self, *args, storage: Storage | str | None = None, **kwargs) -> None:
        try:
            from anthropic import Anthropic as SDKAnthropic
        except ImportError as exc:
            raise RuntimeError("anthropic is required. Install blackboxd[anthropic].") from exc
        super().__init__(SDKAnthropic(*args, **kwargs), storage=storage)


class AsyncAnthropic(_BaseAnthropicWrapper):
    _messages_cls = _AsyncMessagesAPI

    def __init__(self, *args, storage: Storage | str | None = None, **kwargs) -> None:
        try:
            from anthropic import AsyncAnthropic as SDKAsyncAnthropic
        except ImportError as exc:
            raise RuntimeError("anthropic is required. Install blackboxd[anthropic].") from exc
        super().__init__(SDKAsyncAnthropic(*args, **kwargs), storage=storage)

File: typescript/src/providers/anthropic.ts
import type { Storage } from "../storage/base.ts";
import { captureLlmCall } from "../capture.ts";
import type { TokenUsage } from "../types.ts";
import { utcNowIso } from "../utils.ts";

function extractUsage(response: unknown): TokenUsage | undefined {
  if (!response || typeof response !== "object") {
    return undefined;
  }

  const usage = (response as { usage?: Record<string, unknown> }).usage;
  if (!usage) {
    return undefined;
  }

  return {
    inputTokens: typeof usage.input_tokens === "number" ? usage.input_tokens : undefined,
    outputTokens: typeof usage.output_tokens === "number" ? usage.output_tokens : undefined,
  };
}

export class AnthropicWrapper {
  client: {
    messages: { create: (input: Record<string, unknown>) => Promise<unknown> };
  };
  provider: string;
  storage?: Storage | string;

  constructor(
    client: {
      messages: { create: (input: Record<string, unknown>) => Promise<unknown> };
    },
    options: { provider?: string; storage?: Storage | string } = {},
  ) {
    this.client = client;
    this.provider = options.provider ?? "anthropic";
    this.storage = options.storage;
  }

  messages = {
    create: async (input: Record<string, unknown>): Promise<unknown> => {
      const startedAt = utcNowIso();
      try {
        const response = await this.client.messages.create(input);
        await captureLlmCall({
          name: "anthropic.messages.create",
          provider: this.provider,
          model: typeof input.model === "string" ? input.model : undefined,
          prompt: input,
          response,
          startedAt,
          endedAt: utcNowIso(),
          usage: extractUsage(response),
          storage: this.storage,
        });
        return response;
      } catch (error) {
        await captureLlmCall({
          name: "anthropic.messages.create",
          provider: this.provider,
          model: typeof input.model === "string" ? input.model : undefined,
          prompt: input,
          startedAt,
          endedAt: utcNowIso(),
          error,
          storage: this.storage,
        });
        throw error;
      }
    },
  };
}
